buffer/src/main/java/io/netty/buffer/api/MemoryManager.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer.api;
+
+import io.netty.buffer.api.internal.MemoryManagerLoader;
+import io.netty.buffer.api.internal.MemoryManagerOverride;
+import io.netty.util.internal.UnstableApi;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.lang.ref.Cleaner;
+import java.util.Optional;
+import java.util.ServiceConfigurationError;
+import java.util.ServiceLoader;
+import java.util.ServiceLoader.Provider;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+/**
+ * The choice of {@code MemoryManager} implementation also determines the choice of {@link Buffer} implementation.
+ * It is the MemoryManager that implement memory allocation, and how to wrap the allocated memory in a {@link Buffer}
+ * interface.
+ *
+ * @apiNote This is a low-level, {@linkplain UnstableApi unstable}, API that is used for
+ * {@link BufferAllocator BufferAllocator} implementations to build upon.
+ * The methods in this interface are unsafe, because they can be used to violate the safety guarantees of the
+ * {@link Buffer} API, and potentially also the safety guarantees of the JVM.
+ */
+@UnstableApi
+public interface MemoryManager {
+    /**
+     * Get the default, or currently configured, memory managers instance.
+     * @return A MemoryManagers instance.
+     */
+    static MemoryManager instance() {
+        return MemoryManagerOverride.configuredOrDefaultManager();
+    }
+
+    /**
+     * Temporarily override the default configured memory managers instance.
+     * <p>
+     * Calls to {@link #instance()} from within the given supplier will get the given managers instance.
+     *
+     * @param managers Override the default configured managers instance with this instance.
+     * @param supplier The supplier function to be called while the override is in place.
+     * @param <T> The result type from the supplier.
+     * @return The result from the supplier.
+     */
+    static <T> T using(MemoryManager managers, Supplier<T> supplier) {
```

Comment:
It needs to be sufficiently available that we can use it to test the memory segment implementation in the incubating repository. It'd be possible to restrict it with Java modules, but otherwise we can only move it to an internal package.